api/models/v0.3/modelBarcode.js

```diff
@@ -1,65 +1,241 @@
 /* eslint-disable */
+const BarcodeDb = require("../../../db");
+const luhn = require("../../helpers/luhnValidations");
 
 /**
  * Creates Barcode objects.
- * TODO: This might require a DB connection to store and keep track of
- * barcodes being used.
  */
 class Barcode {
-  constructor() {
-    // validate uniqueness of barcodes
-    // get an unused barcode starting in descending order
-    //  { where(used: false).order(barcode: :desc) }
-    const available = "Barcode db object";
-    this.available = !!available;
+  constructor(args) {
+    this.ilsClient = args["ilsClient"];
+    // This will return the instance of the class that's
+    // already connected to the database.
+    this.db = BarcodeDb();
   }
 
-  // update in db that barcode was used
-  markUsed(barcode) {
-    noop();
-  }
-  // this wasnt used in previous codebase
-  markUnused(barcode) {
-    noop();
+  /**
+   * getNextAvailableBarcode()
+   * Generates a barcode in the database from the last one in the database,
+   * and verifies it's available in the ILS. If it is, return it.
+   */
+  async getNextAvailableBarcode() {
+    const { barcode, newBarcode } = await this.nextAvailableFromDB();
+    // Perhaps this barcode from the database isn't actually available in the
+    // ILS. If that's the case, the next available barcode in the ILS is
+    // returned. Otherwise, the initial barcode is the "finalBarcode".
+    const { available, finalBarcode } = await this.availableInIls(
+      barcode,
+      newBarcode
+    );
+
+    // TODO: Figure out how to close a db connection if the service will
+    // be a lambda.
+    // this.release();
+
+    return available ? finalBarcode : undefined;
   }
-  // update in db that barcode is unused
 
-  // Might not need to implement this
-  sendBarcodeAlertEmail(count) {
-    if (Barcode.LOW_BARCODE_ALERT_COUNTS.includes(count)) {
-      // get last barcode from db
-      // send email - AdminMailer was used before
+  /**
+   * nextLuhnValidCode(barcode)
+   * Return the next Luhn-valid barcode.
+   *
+   * @param {string} barcode
+   */
+  nextLuhnValidCode(barcode) {
+    if (!barcode || barcode.length !== 14) {
+      return;
     }
+
+    // Remove the last digit from the existing barcode since we need a
+    // 13-digit number to pass through the luhn-algorithm.
+    let newValidBarcode = Math.floor(parseInt(barcode, 10) / 10);
```

Comment:
It doesn't matter for our purposes, but if a barcode were to start with zeroes, this would strip the zeroes and we'd end up with a barcode smaller than 14 digits.